### PR TITLE
stubtest: analyze `metaclass` of types, refs #13327

### LINKED ISSUE #13327: stubtest: metaclass is not taken into an account
Here's a simple example of this bug:

```python
# runtime implementation
class A(type):
    def method(self) -> None:
        print('a')

class B(metaclass=A): ...

# This works:
# B.method()
```

And:

```python
# type stub
class A(type):
    def method(self) -> None: ...

class B: ...  # notice that `metaclass=A` is missing, but `stubtest` does not produce any errors
```

Notice that this bug prevents using things like `B.method()` in your code.
I will send a fix asap.

Refs https://github.com/python/mypy/issues/13316

### PR DESCRIPTION
**Pull request by sobolevn**
Current `typeshed` produces:

```
error: ast.Bytes metaclass missmatch
Stub: at line 16
Missing metaclass
Runtime: at line 513 in file /Users/sobolev/.pyenv/versions/3.8.9/lib/python3.8/ast.py
Exiting metaclass: <class 'ast._ABC'>

error: ast.Ellipsis metaclass missmatch
Stub: at line 22
Missing metaclass
Runtime: at line 520 in file /Users/sobolev/.pyenv/versions/3.8.9/lib/python3.8/ast.py
Exiting metaclass: <class 'ast._ABC'>

error: ast.NameConstant metaclass missmatch
Stub: at line 21
Missing metaclass
Runtime: at line 517 in file /Users/sobolev/.pyenv/versions/3.8.9/lib/python3.8/ast.py
Exiting metaclass: <class 'ast._ABC'>

error: ast.Num metaclass missmatch
Stub: at line 8
Missing metaclass
Runtime: at line 505 in file /Users/sobolev/.pyenv/versions/3.8.9/lib/python3.8/ast.py
Exiting metaclass: <class 'ast._ABC'>

error: ast.Str metaclass missmatch
Stub: at line 11
Missing metaclass
Runtime: at line 509 in file /Users/sobolev/.pyenv/versions/3.8.9/lib/python3.8/ast.py
Exiting metaclass: <class 'ast._ABC'>

error: contextlib.AsyncExitStack metaclass missmatch
Stub: at line 151
Missing metaclass
Runtime: at line 537 in file /Users/sobolev/.pyenv/versions/3.8.9/lib/python3.8/contextlib.py
Exiting metaclass: <class 'abc.ABCMeta'>

error: contextlib.ExitStack metaclass missmatch
Stub: at line 136
Missing metaclass
Runtime: at line 468 in file /Users/sobolev/.pyenv/versions/3.8.9/lib/python3.8/contextlib.py
Exiting metaclass: <class 'abc.ABCMeta'>

error: dataclasses.InitVar metaclass missmatch
Stub: at line 220
Missing metaclass
Runtime: at line 206 in file /Users/sobolev/.pyenv/versions/3.8.9/lib/python3.8/dataclasses.py
Exiting metaclass: <class 'dataclasses._InitVarMeta'>

error: string.Template metaclass missmatch
Stub: at line 35
Missing metaclass
Runtime: at line 80 in file /Users/sobolev/.pyenv/versions/3.8.9/lib/python3.8/string.py
Exiting metaclass: <class 'string._TemplateMetaclass'>
```

There are two different problems:
1. There are protected metaclasses that do not exist, like `string._TemplateMetaclass`, because we mostly don't add protected names to stubs. This can be solved by adding `and not is_probably_private(stub.metaclass_type.type.fullname)` check (level: primitive). I just wanted to show the full picture: maybe we do want to add them to stubs?
2. `ExitStack`: https://github.com/python/typeshed/issues/7961 This is a real issue with these classes. We can at least add `ABCMeta` metaclass to them

CC @srittau and @AlexWaygood 
Closes #13327

### CODE REVIEW

**Comment by srittau** [comment]
While we usually don't add private elements to stubs, I think we should make an exception for base and metaclasses. Having a correct mro seems more important to me than omitting those classes.

File: mypy/stubtest.py [hunk 1]
```diff
@@ -375,6 +375,40 @@ class SubClass(runtime):  # type: ignore
...
+    # Check metaclass. We exclude protocols, because of how complex
+    # their implementation is in different versions of python.
+    # Enums are also hard, ignoring.
+    if not stub.is_protocol and not stub.is_enum:
+        runtime_metaclass = type(runtime)
+        if runtime_metaclass is not type and stub.metaclass_type is None:
+            # This means that runtime has a custom metaclass, but a stub does not.
+            yield Error(
+                object_path,
+                "metaclass missmatch",
```
Comment (srittau) [inline mypy/stubtest.py:387]:
```suggestion
                "metaclass mismatch",
```

File: mypy/stubtest.py [hunk 2]
```diff
@@ -375,6 +375,40 @@ class SubClass(runtime):  # type: ignore
...
+            # We ignore extra `ABCMeta` metaclass on stubs, this might be typing hack.
+            # We also ignore `builtins.type` metaclass as an implementation detail in mypy.
+            and not mypy.types.is_named_instance(
+                stub.metaclass_type, ("abc.ABCMeta", "builtins.type")
+            )
+        ):
+            # This means that our stub has a metaclass that is not present at runtime.
+            yield Error(
+                object_path,
+                "metaclass missmatch",
```
Comment (srittau) [inline mypy/stubtest.py:405]:
```suggestion
                "metaclass mismatch",
```

**Comment by AlexWaygood** [comment]
I agree with @srittau — I'm fine with adding private classes to stubs if they're used by public classes.

**Comment by sobolevn** [comment]
I want to merge this, any feedback, @hauntsaninja or @JelleZijlstra? 🙂

**Review by hauntsaninja — APPROVED** [review]
Looks great, thank you!

**Comment by sobolevn** [comment]
Thanks everyone!

**Pull request merged**